### PR TITLE
[DECODER] Add Pose Estimation Decoder - @open sesame 6/28 13:29

### PR DESCRIPTION
**Pull request by jijoongmoon**
# PR Description

In order to display the result from pose estimaition, we may need
decoder. The output format of pose estimation is [14, o_Width,
o_Height, 1]. 14 means number of joint of human bone structure, top,
neck, r_shoulder, r_elbow, r_wrist, l_shoulder, l_elbow, l_wrist,
r_hip, r_knee, r_ankle, l_hip, l_knee, l_ankle.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+ *
+ * @see         https://github.com/nnsuite/nnstreamer
+ * @author      Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug         No known bugs except for NYI items
+ *
+ * option1: Video Output Dimension (WIDTH:HEIGHT)
+ * option2: Input Dimension (WIDTH:HEIGHT)
+ */
+
+#ifndef _GNU_SOURCE
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-pose.c]: plz remove _GNU_SOURCE, it is unnecessary.

```
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
```

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+#define _GNU_SOURCE
+#endif
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+#include <glib.h>
+#include "font.h"
+#include <math.h>
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-pose.c]: remove dependency math lib, it is unnecessary.
```
#include <math.h>
```

File: ext/nnstreamer/tensor_decoder/meson.build [hunk 3]
```diff
@@ -66,3 +66,26 @@ static_library('nnstreamer_decoder_bounding_boxes',
...
+]
+
+nnstreamer_decoder_pose_estimation_sources = []
+foreach s : decoder_sub_pose_estimation_sources
+  nnstreamer_decoder_pose_estimation_sources += join_paths(meson.current_source_dir(), s)
+endforeach
+
+shared_library('nnstreamer_decoder_pose_estimation',
+  nnstreamer_decoder_pose_estimation_sources,
+  dependencies: [nnstreamer_dep, glib_dep, gst_dep, libm_dep],
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/meson.build]: remove dependency math lib, it is unnecessary.
```
libm_dep
```

File: ext/nnstreamer/tensor_decoder/meson.build [hunk 4]
```diff
@@ -66,3 +66,26 @@ static_library('nnstreamer_decoder_bounding_boxes',
...
+  install: true,
+  install_dir: decoder_subplugin_install_dir
+)
+static_library('nnstreamer_decoder_pose_estimation',
+  nnstreamer_decoder_pose_estimation_sources,
+  dependencies: [nnstreamer_dep, glib_dep, gst_dep, libm_dep],
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/meson.build]: remove dependency math lib, it is unnecessary.
```
libm_dep
```

**Comment by jaeyun-jung** [comment]
For android build, please add definition for pose-estimation.

In 'jni/nnstreamer.mk'
```
# decoder pose estimation
NNSTREAMER_DECODER_PE_SRCS := \
    $(NNSTREAMER_EXT_HOME)/tensor_decoder/tensordec-pose.c
```

**Comment by jijoongmoon** [comment]
Talking about Android build, there is no build instruction for the tensor decoder for android for just now. So we could add it later in another PR.

File: ext/nnstreamer/tensor_decoder/meson.build (hunk 4 above)
Comment (jijoongmoon) [reply ext/nnstreamer/tensor_decoder/meson.build]: Done.

File: ext/nnstreamer/tensor_decoder/meson.build (hunk 3 above)
Comment (jijoongmoon) [reply ext/nnstreamer/tensor_decoder/meson.build]: Done.

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c (hunk 2 above)
Comment (jijoongmoon) [reply ext/nnstreamer/tensor_decoder/tensordec-pose.c]: Done.

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c (hunk 1 above)
Comment (jijoongmoon) [reply ext/nnstreamer/tensor_decoder/tensordec-pose.c]: Done.

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c [hunk 5]
```diff
@@ -0,0 +1,522 @@
+/**
+ * GStreamer / NNStreamer tensor_decoder subplugin, "Pose estimation"
+ * Copyright (C) 2018 Samsung Electronics Co. Ltd.
+ * Copyright (C) 2018 Jijoong Moon <jijoong.moon@samsung.com>
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_decoder/tensordec-pose.c]: ```s/2018/2019/```  ^^

**Comment by myungjoo** [comment]
LGTM!

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+      pose_init,.exit = pose_exit,.setOption = pose_setOption,.getOutCaps =
+      pose_getOutCaps,.getTransformSize = pose_getTransformSize,.decode =
+      pose_decode,
+};
+
+/** @brief Initialize this object for tensordec-plugin */
+void
+init_pose (void)
+{
+  tensordec_probe (&poseEstimation);
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-pose.c]: please do rebase and fix this.

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+  gst_memory_unmap (out_mem, &out_info);
+  if (gst_buffer_get_size (outbuf) == 0)
+    gst_buffer_append_memory (outbuf, out_mem);
+  return GST_FLOW_OK;
+}
+
+static gchar decoder_subplugin_pose_estimation[] = "pose_estimation";
+/** @brief Pose Estimation tensordec-plugin TensorDecDef instance */
+static TensorDecDef poseEstimation = {
+  .modename = decoder_subplugin_pose_estimation,.type = OUTPUT_VIDEO,.init =
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-pose.c]: now struct for decoder-sub param is changed, need to fix.

Comment (myungjoo) [reply ext/nnstreamer/tensor_decoder/tensordec-pose.c]: is this updated?

File: jni/nnstreamer.mk [hunk 8]
```diff
@@ -60,6 +60,10 @@ NNSTREAMER_DECODER_DV_SRCS := \
 NNSTREAMER_DECODER_IL_SRCS := \
     $(NNSTREAMER_EXT_HOME)/tensor_decoder/tensordec-imagelabel.c
 
+# decoder pose estimation
```
Comment (leemgs) [inline jni/nnstreamer.mk]: * Discussion Question:
In the past, we added the "./jni" folder as a basic set to provide how to make JNI interface with nnstreamer for Android platform. do we have to continually keep nnstreamer/jni/ and nnstreamer-example/android/..../jni/?

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+/** @brief Initialize this object for tensordec-plugin */
+void
+init_pose (void)
+{
+  tensordec_probe (&poseEstimation);
+}
+
+/** @brief Destruct this object for tensordec-plugin */
+void
+fini_pose (void)
```
Comment (leemgs) [inline ext/nnstreamer/tensor_decoder/tensordec-pose.c]: which better for readability?
1. fini_pose
2. finish_pose 
3. and so on

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+  g_array_free (results, TRUE);
+  gst_memory_unmap (out_mem, &out_info);
+  if (gst_buffer_get_size (outbuf) == 0)
+    gst_buffer_append_memory (outbuf, out_mem);
+  return GST_FLOW_OK;
+}
+
+static gchar decoder_subplugin_pose_estimation[] = "pose_estimation";
+/** @brief Pose Estimation tensordec-plugin TensorDecDef instance */
+static TensorDecDef poseEstimation = {
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_decoder/tensordec-pose.c]: please do rebase and fix build error

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c (hunk 9 above)
Comment (jijoongmoon) [reply ext/nnstreamer/tensor_decoder/tensordec-pose.c]: 2 is better. I will change it. Thanks.

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c (hunk 7 above)
Comment (jijoongmoon) [reply ext/nnstreamer/tensor_decoder/tensordec-pose.c]: I will update.

File: ext/nnstreamer/tensor_decoder/tensordec-pose.c (hunk 10 above)
Comment (jijoongmoon) [reply ext/nnstreamer/tensor_decoder/tensordec-pose.c]: I will.

File: jni/nnstreamer.mk (hunk 8 above)
Comment (jijoongmoon) [reply jni/nnstreamer.mk]: I think we have to if we want nnstreamer to be the shared lib.

**Pull request merged**